### PR TITLE
[SPARK-19824][Core] Update JsonProtocol to keep consistent with the UI

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Fix any inconsistent part in JsonProtocol with the UI.
This PR also contains the modifications in #17181 

## How was this patch tested?

Updated JsonProtocolSuite.

Before this change, localhost:8080/json shows:
```
{
  "url" : "spark://xingbos-MBP.local:7077",
  "workers" : [ {
    "id" : "worker-20170615172946-192.168.0.101-49450",
    "host" : "192.168.0.101",
    "port" : 49450,
    "webuiaddress" : "http://192.168.0.101:8081",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497519481722
  }, {
    "id" : "worker-20170615172948-192.168.0.101-49452",
    "host" : "192.168.0.101",
    "port" : 49452,
    "webuiaddress" : "http://192.168.0.101:8082",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497519484160
  }, {
    "id" : "worker-20170615172951-192.168.0.101-49469",
    "host" : "192.168.0.101",
    "port" : 49469,
    "webuiaddress" : "http://192.168.0.101:8083",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497519486905
  } ],
  "cores" : 24,
  "coresused" : 24,
  "memory" : 46080,
  "memoryused" : 3072,
  "activeapps" : [ {
    "starttime" : 1497519426990,
    "id" : "app-20170615173706-0001",
    "name" : "Spark shell",
    "user" : "xingbojiang",
    "memoryperslave" : 1024,
    "submitdate" : "Thu Jun 15 17:37:06 CST 2017",
    "state" : "RUNNING",
    "duration" : 65362
  } ],
  "completedapps" : [ {
    "starttime" : 1497519250893,
    "id" : "app-20170615173410-0000",
    "name" : "Spark shell",
    "user" : "xingbojiang",
    "memoryperslave" : 1024,
    "submitdate" : "Thu Jun 15 17:34:10 CST 2017",
    "state" : "FINISHED",
    "duration" : 116895
  } ],
  "activedrivers" : [ ],
  "status" : "ALIVE"
}
```

After the change:
```
{
  "url" : "spark://xingbos-MBP.local:7077",
  "workers" : [ {
    "id" : "worker-20170615175032-192.168.0.101-49951",
    "host" : "192.168.0.101",
    "port" : 49951,
    "webuiaddress" : "http://192.168.0.101:8081",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497520292900
  }, {
    "id" : "worker-20170615175034-192.168.0.101-49953",
    "host" : "192.168.0.101",
    "port" : 49953,
    "webuiaddress" : "http://192.168.0.101:8082",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497520280301
  }, {
    "id" : "worker-20170615175037-192.168.0.101-49955",
    "host" : "192.168.0.101",
    "port" : 49955,
    "webuiaddress" : "http://192.168.0.101:8083",
    "cores" : 8,
    "coresused" : 8,
    "coresfree" : 0,
    "memory" : 15360,
    "memoryused" : 1024,
    "memoryfree" : 14336,
    "state" : "ALIVE",
    "lastheartbeat" : 1497520282884
  } ],
  "aliveworkers" : 3,
  "cores" : 24,
  "coresused" : 24,
  "memory" : 46080,
  "memoryused" : 3072,
  "activeapps" : [ {
    "id" : "app-20170615175122-0001",
    "starttime" : 1497520282115,
    "name" : "Spark shell",
    "cores" : 24,
    "user" : "xingbojiang",
    "memoryperslave" : 1024,
    "submitdate" : "Thu Jun 15 17:51:22 CST 2017",
    "state" : "RUNNING",
    "duration" : 10805
  } ],
  "completedapps" : [ {
    "id" : "app-20170615175058-0000",
    "starttime" : 1497520258766,
    "name" : "Spark shell",
    "cores" : 24,
    "user" : "xingbojiang",
    "memoryperslave" : 1024,
    "submitdate" : "Thu Jun 15 17:50:58 CST 2017",
    "state" : "FINISHED",
    "duration" : 9876
  } ],
  "activedrivers" : [ ],
  "completeddrivers" : [ ],
  "status" : "ALIVE"
}
```

### CODE REVIEW

**Comment by jiangxb1987** [comment]
cc @yongtang @gatorsmile @cloud-fan

**Comment by gatorsmile** [comment]
Show the screenshot before and after this PR?

**Comment by jiangxb1987** [comment]
@gatorsmile Sure, let me do it later!

**Comment by jiangxb1987** [comment]
retest please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 1]
```diff
@@ -54,7 +55,7 @@ private[deploy] object JsonProtocol {
 
   def writeApplicationDescription(obj: ApplicationDescription): JObject = {
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: IMO, we need to add the function descriptions for these public functions and provide a clear definition for each field. How about adding a few JIRAs and ask the open source community to do it?

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: Let me do this at the weekend in this PR.

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 2]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
-   ("coresused" -> obj.coresUsed) ~
-   ("coresfree" -> obj.coresFree) ~
-   ("memory" -> obj.memory) ~
-   ("memoryused" -> obj.memoryUsed) ~
-   ("memoryfree" -> obj.memoryFree) ~
-   ("state" -> obj.state.toString) ~
-   ("lastheartbeat" -> obj.lastHeartbeat)
- }
+  /**
+   * Export the [[WorkerInfo]] to a Json object, a [[WorkerInfo]] consists of the information of a
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, a` -> `. A`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 3]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   * worker.
+   *
+   * @return a Json object containing the following fields:
+   *         `id` a string identifier of the worker
+   *         `host` the host that the worker is running on
+   *         `port` the port that the worker is bound to
+   *         `address` ${host}:${port}
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: If we already has host and port, why we still need the `address`?

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 4]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+    ("lastheartbeat" -> obj.lastHeartbeat)
+  }
 
+  /**
+   * Export the [[ApplicationInfo]] to a Json object, an [[ApplicationInfo]] consists of the
+   * information of an application.
+   *
+   * @return a Json object containing the following fields:
+   *         `id` a string identifier of the application
+   *         `starttime` time in milliseconds that the application starts
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: Switch the order?

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 5]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         `name` a name describes the application
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: -> the description of the application

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 6]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+    ("coresfree" -> obj.coresFree) ~
+    ("memory" -> obj.memory) ~
+    ("memoryused" -> obj.memoryUsed) ~
+    ("memoryfree" -> obj.memoryFree) ~
+    ("state" -> obj.state.toString) ~
+    ("lastheartbeat" -> obj.lastHeartbeat)
+  }
 
+  /**
+   * Export the [[ApplicationInfo]] to a Json object, an [[ApplicationInfo]] consists of the
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, an` -> `. An`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 7]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+    ("cores" -> obj.coresGranted) ~
     ("user" -> obj.desc.user) ~
     ("memoryperslave" -> obj.desc.memoryPerExecutorMB) ~
     ("submitdate" -> obj.submitDate.toString) ~
     ("state" -> obj.state.toString) ~
     ("duration" -> obj.duration)
   }
 
+  /**
+   * Export the [[ApplicationDescription]] to a Json object, an [[ApplicationDescription]] consists
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, an` -> `. An`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 8]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   * of the description of an application.
+   *
+   * @return a Json object containing the following fields:
+   *         `name` a name describes the application
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: -> the description of the application

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 9]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         `cores` max cores can be allocated to the application, 0 means unlimited
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `max cores can be ` -> `max cores that can be  `

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 10]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         `memoryperslave` minimal memory in MB required to each executor
+   *         `user` name of the user who submitted the application
+   *         `command` the command string that submitted the application
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `the command string that submitted` -> `the command submitted`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 11]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
     ("name" -> obj.name) ~
-    ("cores" -> obj.maxCores) ~
+    ("cores" -> obj.maxCores.getOrElse(0)) ~
     ("memoryperslave" -> obj.memoryPerExecutorMB) ~
     ("user" -> obj.user) ~
     ("command" -> obj.command.toString)
   }
 
+  /**
+   * Export the [[ExecutorRunner]] to a Json object, an [[ExecutorRunner]] consists of the
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, an` -> `. An`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 12]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   * information of an executor.
+   *
+   * @return a Json object containing the following fields:
+   *         `id` a integer identifier of the executor
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `a integer` -> `an integer`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 13]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         `memory` memory in MB allocated to the executor
+   *         `appid` a string identifier of the application that the executor is working for
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `working for` -> `working on`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 14]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         `appdesc` a Json object of the [[ApplicationDescription]] of the application that the
+   *         executor is working for
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `working for` -> `working on`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 15]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   */
   def writeExecutorRunner(obj: ExecutorRunner): JObject = {
     ("id" -> obj.execId) ~
     ("memory" -> obj.memory) ~
     ("appid" -> obj.appId) ~
     ("appdesc" -> writeApplicationDescription(obj.appDesc))
   }
 
+  /**
+   * Export the [[DriverInfo]] to a Json object, a [[DriverInfo]] consists of the information of a
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, a` -> `. A`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 16]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
     ("cores" -> obj.desc.cores) ~
-    ("memory" -> obj.desc.mem)
+    ("memory" -> obj.desc.mem) ~
+    ("submitdate" -> obj.submitDate.toString) ~
+    ("worker" -> obj.worker.map(_.id).getOrElse("None")) ~
+    ("mainclass" -> obj.desc.command.arguments(2))
   }
 
+  /**
+   * Export the [[MasterStateResponse]] to a Json object, a [[MasterStateResponse]] consists the
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: `, a` -> `. A`

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala [hunk 17]
```diff
@@ -21,74 +21,192 @@ import org.json4s.JsonAST.JObject
...
+   *         master
+   *         `aliveworkers` size of alive workers allocated to the master
+   *         `cores` total cores available of the master
+   *         `coresused` cores used by the master
+   *         `memory` total memory available of the master
+   *         `memoryused` memory used by the master
+   *         `activeapps` a list of Json objects of [[ApplicationInfo]] of the active applications
+   *         running on the master
+   *         `completedapps` a list of Json objects of [[ApplicationInfo]] of the completed
+   *         applications from the master
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: -> `the applications completed in the master`

**Comment by gatorsmile** [comment]
LGTM except a few comments.

File: core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala (hunk 3 above)
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/JsonProtocol.scala]: Good point, I'll remove this.

**Comment by gatorsmile** [comment]
LGTM

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**